journalbeat/checkpoint/checkpoint.go

```diff
@@ -0,0 +1,290 @@
+	maxUpdates    int            // Maximum number of updates to buffer before persisting to disk.
+	flushInterval time.Duration  // Maximum time interval that can pass before persisting to disk.
+	sort          []string       // Slice used for sorting states map (store to save on mallocs).
+
+	lock   sync.RWMutex
+	states map[string]JournalState
+
+	save chan JournalState
+}
+
+// PersistedState represents the format of the data persisted to disk.
+type PersistedState struct {
+	UpdateTime time.Time      `yaml:"update_time"`
+	States     []JournalState `yaml:"journal_entries"`
+}
+
+// JournalState represents the state of an individual event log.
+type JournalState struct {
+	Path               string `yaml:"path"`
+	Cursor             string `yaml:"cursor"`
+	RealtimeTimestamp  uint64 `yaml:"realtime_timestamp"`
+	MonotonicTimestamp uint64 `yaml:"monotonic_timestamp"`
+}
+
+// NewCheckpoint creates and returns a new Checkpoint. This method loads state
+// information from disk if it exists and starts a goroutine for persisting
+// state information to disk. Shutdown should be called when finished to
+// guarantee any in-memory state information is flushed to disk.
+//
+// file is the name of the file where event log state is persisted as YAML.
+// maxUpdates is the maximum number of updates checkpoint will accept before
+// triggering a flush to disk. interval is maximum amount of time that can
+// pass since the last flush before triggering a flush to disk (minimum value
+// is 1s).
+func NewCheckpoint(file string, maxUpdates int, interval time.Duration) (*Checkpoint, error) {
+	c := &Checkpoint{
+		done:          make(chan struct{}),
+		file:          file,
+		maxUpdates:    maxUpdates,
+		flushInterval: interval,
+		sort:          make([]string, 0, 10),
+		states:        make(map[string]JournalState),
+		save:          make(chan JournalState, 1),
+	}
+
+	// Minimum batch size.
+	if c.maxUpdates < 1 {
+		c.maxUpdates = 1
+	}
+
+	// Minimum flush interval.
+	if c.flushInterval < time.Second {
+		c.flushInterval = time.Second
+	}
+
+	// Read existing state information:
+	ps, err := c.read()
+	if err != nil {
+		return nil, err
+	}
+
+	if ps != nil {
+		for _, state := range ps.States {
+			c.states[state.Path] = state
+		}
+	}
+
+	// Write the state file to verify we have have permissions.
+	err = c.flush()
+	if err != nil {
+		return nil, err
+	}
+
+	c.wg.Add(1)
+	go c.run()
+	return c, nil
+}
+
+// run is worker loop that reads incoming state information from the save
+// channel and persists it when the number of changes reaches maxEvents or
+// the amount of time since the last disk write reaches flushInterval.
+func (c *Checkpoint) run() {
+	defer c.wg.Done()
+	defer c.persist()
+
+	flushTimer := time.NewTimer(c.flushInterval)
+	defer flushTimer.Stop()
+loop:
+	for {
+		select {
+		case <-c.done:
+			break loop
+		case s := <-c.save:
+			c.lock.Lock()
+			c.states[s.Path] = s
```

Comment:
Instead of using locks around c.states could we use some methods like `setState` that have the lock inside. Like this we make sure we never forget.

Reply:
I just simply copied checkpointing from Winlogbeat and haven't touched it apart from adjusting the structs which are stored. I would rather not make further changes, because we are going to port everything to the new registry soon.

Reply:
Got it. @andrewkroh FYI